DSTA-302: add_reporting_period.py + file structure, poetry, readme (#28)
NHSDigital/dtos-analyse-data-science: projects/kc63_preprocessing/README.md

# KC63 Preprocessing

A Python toolkit for preprocessing KC63 data by adding reporting period information to CSV files.

## Overview

This toolkit processes KC63 CSV files and adds a reporting period column based on the year extracted from the filename. The reporting period follows the format `{year-1}-{year}`, representing the fiscal year.

Key features:

- Extracts year from filename (expects files ending with a 4-digit year)

> - Automatically generates reporting period in the format `YYYY-YYYY`

Reviewer: Nit-pick: I think this is clearer for the reader.

```suggestion
- Automatically generates reporting period in the format `{YYYY-1}-{YYYY}`
```

Author: Updated

- Batch processes all CSV files in an input directory

> - Preserves original data while adding the new reporting period column

Reviewer: Nit-pick: Is this more accurate:

```suggestion
- Copies the original data to a new file and adds a new `reporting_period` column
```

Author: Updated

## Installation

### Prerequisites

> - Python 3.8 or higher
> - Required Python packages:
>   - `pandas`
> 
> ### Setup
> 
> 1. Install the requirements:
> 
> ```bash
> poetry install
> ```

Reviewer: Nit-pick: I think we assume it's implicit that the required packages will be installed so we don't need to maintain two lists. We could add a comment that 'this will install the dependencies listed in poetry.lock' to help the reader.

```suggestion
### Prerequisites

- Python 3.8 or higher

### Setup

1. Install the requirements:

```bash
poetry install
```
```

Author: Updated

> This will install the requirements in the `poetry.lock` file

Reviewer: Nit-pick: full stop

```suggestion
This will install the requirements in the `poetry.lock` file.
```

## Usage

### Add Reporting Period to CSV Files

The script processes CSV files located in the `reporting_period/input/` directory and outputs the results to `reporting_period/output/`.

1. Place your CSV files in the `reporting_period/input/` directory. Filenames must end with a 4-digit year (e.g., `data_2023.csv`).

2. Run the script:

```bash
cd reporting_period
poetry run python add_reporting_period.py
```

The script will:

- Read each CSV file from the input directory
- Extract the year from the filename

> - Add a `reporting_period` column with the value `{year-1}-{year}`

Reviewer: Suggestion: use either `{year-1}-{year}` or `{YYYY-1}-{YYYY}` throughout documentation. I think `{YYYY-1}-{YYYY}` is more common.

Author: Updated to {YYYY-1}-{YYYY}

> - Save the processed file to the output directory with the same filename

Reviewer: Suggestion: add a suffix to the output filename to avoid the possibility of confusing files with the same name but different content.

Author: Updated

### Example

For a file named `kc63_data_2023.csv`, the script will:

- Extract year: `2023`
- Generate reporting period: `2022-2023`

> - Add this value to a new `reporting_period` column in the CSV

Reviewer: Clarification:

```suggestion
- Extract year: `2023`
- Generate reporting period value: `2022-2023`
- Add this value to a new `reporting_period` column in the outgoing CSV
```

Author: Updated

poetry install
```

> This will install the requirements in the `poetry.lock` file.

Reviewer: ❤️
